### PR TITLE
[Breaking] Migrate to scala 3

### PR DESCRIPTION
**Pull request by RemiBardon**

### CODE REVIEW

**Comment by RemiBardon** [comment]
CI has passed successfully 🥳 However I don't think CI builds for 3.2.0 yet 🥲

@judu Should I change the CI build versions?

**Comment by RemiBardon** [comment]
As said on Slack, I had to duplicate a test suite for Scala 2 and Scala 3 because Circe doesn't give the same results in those two versions 🤦🏻‍♂️

I'm marking this PR as ready, but we can fix the test issue before merging if you want.

**Comment by RemiBardon** [comment]
(If someone else merges this PR: please squash merge)

**Review by judu — COMMENTED** [review]
Needs a full documentation of the release when releasing.

File: build.sbt [hunk 1]
```diff
@@ -50,7 +56,16 @@ lazy val commonSettings = Seq(
   Global / parallelExecution := false,
   Global / concurrentRestrictions += Tags.limit(Tags.Test, 1),
   Compile / doc / scalacOptions := (Compile / doc / scalacOptions).value.filter(_ != "-Xfatal-warnings"),
-  scalacOptions ++= Seq("-unchecked", "-deprecation", "-encoding", "utf8")
+  scalacOptions ++= Seq("-unchecked", "-encoding", "utf8")
+    ++ (CrossVersion.partialVersion(scalaVersion.value) match {
+      case Some((3, _)) => Seq(
+        "-source:3.0-migration",
```
Comment (judu) [inline build.sbt:62]:
```suggestion
```

[the remark this answers is not in the document]
Comment (RemiBardon) [reply build.sbt:62]: Oh yeah I believe so

File: build.sbt [hunk 2]
```diff
@@ -50,7 +56,16 @@ lazy val commonSettings = Seq(
...
+        "-rewrite",
```
Comment (judu) [inline build.sbt:63]:
```suggestion
```

File: build.sbt [hunk 3]
```diff
@@ -50,7 +56,16 @@ lazy val commonSettings = Seq(
...
+      )
```
Comment (RemiBardon) [inline build.sbt:64]:
```suggestion
```

File: build.sbt [hunk 4]
```diff
@@ -50,7 +56,16 @@ lazy val commonSettings = Seq(
   Global / parallelExecution := false,
   Global / concurrentRestrictions += Tags.limit(Tags.Test, 1),
   Compile / doc / scalacOptions := (Compile / doc / scalacOptions).value.filter(_ != "-Xfatal-warnings"),
-  scalacOptions ++= Seq("-unchecked", "-deprecation", "-encoding", "utf8")
+  scalacOptions ++= Seq("-unchecked", "-encoding", "utf8")
+    ++ (CrossVersion.partialVersion(scalaVersion.value) match {
+      case Some((3, _)) => Seq(
```
Comment (RemiBardon) [inline build.sbt:61]:
```suggestion
      case Some((3, _)) => Seq()
```

File: build.sbt [hunk 5]
```diff
@@ -50,7 +56,13 @@ lazy val commonSettings = Seq(
   Global / parallelExecution := false,
   Global / concurrentRestrictions += Tags.limit(Tags.Test, 1),
   Compile / doc / scalacOptions := (Compile / doc / scalacOptions).value.filter(_ != "-Xfatal-warnings"),
-  scalacOptions ++= Seq("-unchecked", "-deprecation", "-encoding", "utf8")
+  scalacOptions ++= Seq("-unchecked", "-encoding", "utf8")
+    ++ (CrossVersion.partialVersion(scalaVersion.value) match {
+      case Some((3, _)) => Seq()
+      case _ => Seq(
+        "-deprecation",
+      )
```
Comment (RemiBardon) [inline build.sbt:64]:
```suggestion
      case Some((2, _)) => Seq(
        "-deprecation",
      )
      case _ => Seq()
```

**Review by judu — APPROVED** [review]
I've reviewed this with Rémi and then fixed some stuff offline.

**Pull request merged**